src/environment.cpp

```diff
@@ -271,6 +271,210 @@ ABMWithState::ABMWithState(ActiveBlockModifier *abm_):
+		std::vector<std::string> components = str_split(entry, ':');
+		if (components.size() != 2)
+			throw SerializationError("Introduction times entry \""
+				+ entry + "\" requires exactly one ':'!");
+		u64 hash = stou64(components[0]);
+		u32 time = (u32)stoi64(components[1]);
+		introduction_times[hash] = time;
+	}
+
+	// Put stuff from introduction_times into m_lbm_lookup
+	for (std::map<u64, u32>::const_iterator it = introduction_times.begin();
+			it != introduction_times.end(); ++it) {
+		u32 hash = it->first;
+		u32 time = it->second;
+
+		std::map<u32, LoadingBlockModifierDef *>::iterator def_it =
+			m_lbm_defs.find(hash);
+		if (def_it == m_lbm_defs.end()) {
+			// This seems to be an LBM entry for
+			// an LBM we haven't loaded. Discard it.
+			continue;
+		}
+		LoadingBlockModifierDef *lbm_def = def_it->second;
+
+		lbm_lookup_map::iterator jt = m_lbm_lookup.find(time);
+		if (jt == m_lbm_lookup.end()) {
+			LBMContentMapping m;
+			m_lbm_lookup[time] = m;
+			jt = m_lbm_lookup.find(time);
+		}
+		jt->second.addLBM(lbm_def, gamedef);
+
+		// Erase the entry so that we know later
+		// what element didn't get put into m_lbm_lookup
+		m_lbm_defs.erase(hash);
+	}
+
+	// Now also add the elements from m_lbm_defs to m_lbm_lookup
+	// that weren't mentioned by introduction_times.
+	// They are introduced first time to this world.
+	lbm_lookup_map::iterator jt = m_lbm_lookup.find(now);
+	if (jt == m_lbm_lookup.end()) {
+		LBMContentMapping m;
+		m_lbm_lookup[now] = m;
+		jt = m_lbm_lookup.find(now);
+	}
+	LBMContentMapping &lbms_we_introduce_now = jt->second;
+
+	for (std::map<u32, LoadingBlockModifierDef *>::iterator it = m_lbm_defs.begin();
+			it != m_lbm_defs.end(); ++it) {
+		lbms_we_introduce_now.addLBM(it->second, gamedef);
+	}
+
+	// Clear the list, so that we don't delete remaining elements
+	// twice in the destructor
+	m_lbm_defs.clear();
+}
+
+std::string LBMManager::createIntroductionTimesString()
+{
+	// Precondition, we must be in query mode
+	FATAL_ERROR_IF(m_query_mode == false,
+		"attempted to query on non fully set up LBMManager");
+
+	std::ostringstream oss;
+	for (lbm_lookup_map::iterator it = m_lbm_lookup.begin();
+			it != m_lbm_lookup.end(); ++it) {
+		u32 time = it->first;
+		std::vector<LoadingBlockModifierDef *> &lbm_list = it->second.lbm_list;
+		for (std::vector<LoadingBlockModifierDef *>::iterator iit = lbm_list.begin();
+				iit != lbm_list.end(); ++iit) {
+			oss << (*iit)->nameHash << ":" << time << ";";
+		}
+	}
+	return oss.str();
+}
+
+void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+{
+	// Precondition, we need m_lbm_lookup to be initialized
+	FATAL_ERROR_IF(m_query_mode == false,
+		"attempted to query on non fully set up LBMManager");
+	v3s16 pos;
+	for (pos.X = 0; pos.X < MAP_BLOCKSIZE; pos.X++)
+	for (pos.Y = 0; pos.Y < MAP_BLOCKSIZE; pos.Y++)
+	for (pos.Z = 0; pos.Z < MAP_BLOCKSIZE; pos.Z++)
+	{
+		MapNode n = block->getNodeNoEx(pos);
```

Comment:
move the variable declaration outside of the loop to prevent program to create/destroy variable at each step